fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/ModifyInference.java

```diff
@@ -0,0 +1,181 @@
+    }
+
+    @Override
+    public ModifyOp visitPhysicalProject(OptExpression optExpression, Void ctx) {
+        return visit(optExpression.inputAt(0), ctx);
+    }
+
+    // TODO(murphy) read from user property, support custom the ModifyOp behavior
+    private ModifyOp visitOlapTable(Table table) {
+        if (!table.isOlapTable()) {
+            throw new NotImplementedException("Only OLAP table is supported");
+        }
+        OlapTable olapTable = (OlapTable) table;
+        ModifyOp res;
+        if (olapTable.getKeysType().equals(KeysType.DUP_KEYS)) {
+            res = ModifyOp.INSERT_ONLY;
+        } else {
+            res = ModifyOp.ALL;
+        }
+        return res;
+    }
+
+    @Override
+    public ModifyOp visitPhysicalOlapScan(OptExpression optExpression, Void ctx) {
+        PhysicalOlapScanOperator scan = (PhysicalOlapScanOperator) optExpression.getOp();
+        return visitOlapTable(scan.getTable());
+    }
+
+    @Override
+    public ModifyOp visitPhysicalStreamScan(OptExpression optExpression, Void ctx) {
+        PhysicalStreamScanOperator scan = (PhysicalStreamScanOperator) optExpression.getOp();
+        ModifyOp op = visitOlapTable(scan.getTable());
+        scan.setModifyOp(op);
+        return op;
+    }
+
+    @Override
+    public ModifyOp visitPhysicalStreamJoin(OptExpression optExpression, Void ctx) {
+        ModifyOp lhsOp = infer(optExpression.inputAt(0));
+        ModifyOp rhsOp = infer(optExpression.inputAt(1));
+        PhysicalStreamJoinOperator join = (PhysicalStreamJoinOperator) optExpression.getOp();
+
+        // TODO(murphy) support outer join
+        if (!join.getJoinType().isInnerJoin()) {
+            throw new NotImplementedException("Only INNER JOIN is supported");
+        }
+        ModifyOp joinModify = ModifyOp.union(lhsOp, rhsOp);
+        join.setModifyOp(joinModify);
+        return joinModify;
+    }
+
+    @Override
+    public ModifyOp visitPhysicalStreamAgg(OptExpression optExpression, Void ctx) {
+        ModifyOp inputModify = infer(optExpression.inputAt(0));
+        PhysicalStreamAggOperator agg = (PhysicalStreamAggOperator) optExpression.getOp();
+
+        // TODO(murphy) optimize for non-retractable scenario
+        ModifyOp res;
+        if (inputModify.isInsertOnly()) {
+            res = ModifyOp.UPSERT;
+        } else {
+            res = ModifyOp.ALL;
+        }
+        agg.setModifyOp(res);
+        return res;
+    }
+
+    public static class ModifyOp {
+        private final EnumSet<ModifyKind> modifySet;
+        private final EnumSet<UpdateKind> updateSet;
+
+        public static final ModifyOp NONE = new ModifyOp(ModifyKind.NONE, UpdateKind.NONE);
+        public static final ModifyOp INSERT_ONLY = new ModifyOp(ModifyKind.INSERT_ONLY, UpdateKind.NONE);
+        public static final ModifyOp UPSERT = new ModifyOp(ModifyKind.UPSERT, UpdateKind.ONLY_AFTER);
+        public static final ModifyOp ALL = new ModifyOp(ModifyKind.ALL, UpdateKind.ALL);
+
```

Comment:
so What's the ModifyOp ? Different ModifyOps will generate which physical property?  

Is there some information to tell Physical operator which can not generate `retract` messages?

Reply:
Each stream operator has a `ModifyOp`, such as `INSERT_ONLY`, `UPSERT`, `ALL`.

Take the `StreamAgg` Operator as example:
- If its input is `INSERT_ONLY`, the agg itself is `UPSERT`, so it generates `INSERT` and `DELETE` only, but not `UPDATE`
- If its input is `ALL`, the agg would output all kinds of `ModifyOp`